Mathlib/Topology/Separation/DisjointCover.lean

```diff
@@ -58,3 +59,110 @@ lemma exists_finite_nonempty_disjoint_clopen_cover (hU : IsOpenCover U) :
   · exact hWd.comp_of_injective <| Subtype.val_injective.comp t.equivFin.symm.injective
 
 end TopologicalSpace.IsOpenCover
+
+namespace TopologicalSpace
+variable {X : Type*} [TopologicalSpace X] {S : Set (X × X)}
+
+/-- If `S` is any neighbourhood of the diagonal in a topological space `X`, any point of `X` has an
+open neighbourhood `U` such that `U ×ˢ U ⊆ S`. -/
+lemma exists_open_prod_subset_of_mem_nhds_diagonal (hS : S ∈ nhdsSet (diagonal X)) (x : X) :
+    ∃ U : Set X, IsOpen U ∧ x ∈ U ∧ U ×ˢ U ⊆ S := by
+  have : S ∈ 𝓝 (x, x) := mem_nhdsSet_iff_forall.mp hS _ rfl
+  obtain ⟨u, v, huo, hux, hvo, hvx, H⟩ := by rwa [mem_nhds_prod_iff'] at this
+  exact ⟨_, huo.inter hvo, ⟨hux, hvx⟩, fun p hp ↦ H ⟨hp.1.1, hp.2.2⟩⟩
+
+variable [CompactSpace X]
+
+/-- If `S` is any neighbourhood of the diagonal in a compact topological space `X`, then there
+exists a finite cover of `X` by opens `U i` with `U i ×ˢ U i ⊆ S` for all `i`.
+
+That the indexing set is a finset of `X` is an artifact of the proof; it could be any finite type.
+-/
+lemma exists_finite_open_cover_prod_subset_of_mem_nhds_diagonal_of_compact
+    (hS : S ∈ nhdsSet (.diagonal X)) :
+    ∃ (t : Finset X) (U : t → Opens X), IsOpenCover U ∧ ∀ i, (U i : Set X) ×ˢ U i ⊆ S := by
+  choose U hUo hUx hUp using exists_open_prod_subset_of_mem_nhds_diagonal hS
+  obtain ⟨t, ht⟩ := isCompact_univ.elim_finite_subcover _ hUo (fun x _ ↦ mem_iUnion.mpr ⟨_, hUx x⟩)
+  refine ⟨t, fun i ↦ ⟨_, hUo i⟩, .of_sets _ ?_, (hUp ·)⟩
+  simpa [iUnion_subtype, ← univ_subset_iff] using ht
+
+variable [TotallyDisconnectedSpace X] [T2Space X]
+
+/-- If `S` is any neighbourhood of the diagonal in a profinite topological space `X`, then there
+exists a finite cover of `X` by disjoint nonempty clopens `U i` with `U i ×ˢ U i ⊆ S` for all `i`.
+-/
+private lemma exists_finite_disjoint_nonempty_clopen_cover_of_mem_nhds_diagonal_of_profinite
+    (hS : S ∈ nhdsSet (.diagonal X)) :
+    ∃ (n : ℕ) (D : Fin n → Clopens X), (∀ i, D i ≠ ⊥) ∧
+    (∀ i, ∀ y ∈ D i, ∀ z ∈ D i, (y, z) ∈ S) ∧ (univ : Set X) ⊆ ⋃ i, ↑(D i) ∧
+    Pairwise (Disjoint on D) := by
+  obtain ⟨t, U, hUc, hUS⟩ := exists_finite_open_cover_prod_subset_of_mem_nhds_diagonal_of_compact hS
+  -- Now refine it to a disjoint covering.
+  obtain ⟨n, W, hW₁, hW₂, hW₃⟩ := hUc.exists_finite_nonempty_disjoint_clopen_cover
+  refine ⟨n, W, fun j ↦ (hW₁ j).1, fun j y hy z hz ↦ ?_, hW₂, hW₃⟩
+  exact match (hW₁ j).2 with | ⟨i, hi⟩ => hUS i ⟨hi hy, hi hz⟩
+
+variable {V : Type*} [TopologicalSpace V] (f : C(X, V)) {S : Set (V × V)}
+
+/--
+For any continuous function `X → V`, with `X` profinite and `V` Hausdorff, and `S` a
```

Comment:
You don't actually need Hausdorff for the next lemma. Also, I do not understand why it is still in the `TopologicalSpace` namespace and not already in the `ContinuousMap` one.